### ISSUE TITLE
Support Concatenate in ParamSpec aliases

### ISSUE THREAD

**Issue by bijij** [microsoft/pyright#3177]
**The problem**
To simplify some code I have some commonly used annotations shortened via type-aliases, e.g.
```py
Coro = Coroutine[Any, Any, T]
CoroFunc = Callable[P, Coro[T]]
```

I wanted to create an aditional alias which described a generic check function
```py
CheckFunc = CoroFunc[Concatenate[Foo, P], bool]
```

However pyright returns the following error:
`error: Expected ParamSpec, ellipsis, or list of types`

If the inner alias is removed, i.e. CheckFunc is defined as 
```py
CheckFunc = Callable[Concatenate[Foo, P], Coro[bool]]
```
Then my code type-checks as I'd expect.

**Describe the solution you'd like**:
`Concatenate` should be supported in this context, allowing for aliases with pre-set arguments.

And thus:
```py
async def my_check_func(foo: Foo, a: int, b: str) -> bool:
   return str(a) == b

async def takes_check_func(check_func: CheckFunc[P], *args: P.args, **kwargs: P.kwargs):
    await check_func(Foo(), *args, **kwargs)

asyncio.run(takes_check_func(my_check_func, 1, '2'))
```
Code similar to this should type-check

For what it's worth, this code:
```py
from typing import Any, Callable, Coroutine, TypeVar, Union, ParamSpec, Concatenate

T = TypeVar('T')
P = ParamSpec('P')

Coro = Coroutine[Any, Any, T]
MaybeCoro = Union[T, Coro[T]]

CoroFunc = Callable[P, Coro[T]]
MaybeCoroFunc = Union[Callable[P, T], CoroFunc[P, T]]

import asyncio

class Foo:
    ...

CheckFunc = CoroFunc[Concatenate[Foo, P], Coro[bool]]

async def my_check_func(foo: Foo, a: int, b: str) -> bool:
    print(a, b)
    return str(a) == b

async def takes_check_func(check_func: CheckFunc[P], *args: P.args, **kwargs: P.kwargs):
    await check_func(Foo(), *args, **kwargs)


asyncio.run(takes_check_func(my_check_func, 1, '2'))
```

Works without issue at runtime on 3.10.

**Comment by bijij** [comment]
I have found that the issue was a result of pyright assuming the running version was 3.8
which this code does not work on

Modifying it to appear such as this works as expected.
```py
from __future__ import annotations

from typing import Any, Callable, Coroutine, TypeVar
from typing_extensions import ParamSpec, Concatenate, TypeAlias

T = TypeVar('T')
P = ParamSpec('P')

Coro = Coroutine[Any, Any, T]
CoroFunc = Callable[P, Coro[T]]


import asyncio

class Foo:
    ...

CheckFunc: TypeAlias = "Callable[Concatenate[Foo, P], Coro[bool]]"

async def my_check_func(foo: Foo, a: int, b: str) -> bool:
    print(a, b)
    return str(a) == b

async def takes_check_func(check_func: CheckFunc[P], *args: P.args, **kwargs: P.kwargs):
    await check_func(Foo(), *args, **kwargs)


asyncio.run(takes_check_func(my_check_func, 1, '2'))
```

**Comment by bijij** [comment]
Seems I closed this prematurely, 
```py
from typing import Any, Callable, Coroutine, TypeVar, Union, ParamSpec, Concatenate, TypeAlias

T = TypeVar('T')
U = TypeVar('U')
P = ParamSpec('P')


Coro = Coroutine[Any, Any, T]
MaybeCoro = Union[T, Coro[T]]

CoroFunc = Callable[P, Coro[T]]
MaybeCoroFunc = Union[Callable[P, T], CoroFunc[P, T]]


MaybeMethod = Union[
    Callable[Concatenate[T, P], U],
    Callable[P, U]
]

CoroMaybeMethod: TypeAlias = Union[
    CoroFunc[Concatenate[T, P], U],
    CoroFunc[P, U]
]

MaybeCoroMaybeMethod: TypeAlias = Union[
    MaybeCoroFunc[Concatenate[T, P], U],
    MaybeCoroFunc[P, U]
]
```

Not the most practical code probably but this doesn't seem to work at all no-matter what i try,
It's all perfectly acceptable at runtime however, at least on 3.10

**Comment by bijij** [comment]
```py
from typing import Any, Callable, Coroutine, TypeVar, Union, ParamSpec, Concatenate

T = TypeVar('T')
U = TypeVar('U')
P = ParamSpec('P')


Coro = Coroutine[Any, Any, T]
MaybeCoro = Union[T, Coro[T]]

CoroFunc = Callable[P, Coro[T]]
MaybeCoroFunc = Union[Callable[P, T], CoroFunc[P, T]]

Method = Callable[Concatenate[T, P], U]

CoroMethod = Method[T, P, Coro[U]]
MaybeCoroMethod = Union[
    Method[T, P, U],
    CoroMethod[T, P, U]
]

MaybeMethod = Union[
    Method[T, P, U],
    Callable[P, U]
]

CoroMaybeMethod = Union[
    CoroMethod[T, P, U],
    CoroFunc[P, U]
]

MaybeCoroMaybeMethod = Union[
    MaybeCoroMethod[T, P, U],
    MaybeCoroFunc[P, U]
]
```

By adding a extra `Method`, `CoroMethod` and `MaybeCoroMethod` aliases I can get this to type-check, but I'm confused as to why it fails without additional abstraction.

**Comment by bijij** [comment]
Seems I've found a bug:
![image](https://user-images.githubusercontent.com/8677174/158013763-3844e907-de28-49fc-9225-c9a448e5ffba.png)

Bar's definition for some reason relies on Foo

**Comment by erictraut** [comment]
I've added support for the use of `Concatenate` when specifying a type argument for a generic type alias that accepts a ParamSpec. This will be included in the next release. If you see other problems, please open additional bug reports.

**Comment by erictraut** [comment]
This is included in pyright 1.1.230, which I just published. It will also be included in the next release of pylance.
